### PR TITLE
shadow: support metadata

### PR DESCRIPTION
**Pull request by at-wat**
#211

### CODE REVIEW

File: shadow/state.go [hunk 1]
```diff
@@ -121,6 +138,17 @@ func updateStateRaw(state map[string]interface{}, update json.RawMessage) error
...
 
+func updateStateMetadataRaw(state map[string]interface{}, update json.RawMessage) error {
+	if updateStateRawCommon(state, update) {
+		return nil
+	}
+	var u NestedMetadata
+	if err := json.Unmarshal([]byte(update), &u); err != nil {
+		return ioterr.New(err, "unmarshaling update")
+	}
+	return updateState(state, u)
```
Comment (at-wat) [inline shadow/state.go:149]: Need to check attribute removal logic is same as state document or not.

Comment (at-wat) [reply shadow/state.go:149]: Seems working as expected in the updated example da1401c

**Comment by at-wat** [comment]
@pablocastilla could you try this branch?

Timestamp of `"attribute"` delta can be obtained by `document.Metadata.Delta["attribute"].(shadow.Metadata).Timestamp`.
<details>
<summary>example output of examples/shadow</summary>

```shell
$ cd go/src/github.com/seqsense/aws-iot-device-sdk-go/examples/shadow
$ go run . ${AWS_IOT_ENDPOINT} ${THING_NAME} ${SHADOW_NAME}
Active: <nil>
> update desire
delta:
  Value: 	1
  Struct: 	map[Values:[1 2]]
document:
  State: 
    Desired: 	map[Struct:map[Values:[1 2]] Value:1]
    Reported: 	map[]
    Delta: 	map[Struct:map[Values:[1 2]] Value:1]
  Metadata: 
    Desired: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
    Reported: 	map[]
    Delta: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
  Version: 	191
  Timestamp: 	1613810500
  ClientToken: 	

> update report
document:
  State: 
    Desired: 	map[Struct:map[Values:[1 2]] Value:1]
    Reported: 	map[Struct:map[Values:[1 2]] Value:2]
    Delta: 	map[Struct:map[Values:[1 2]] Value:1]
  Metadata: 
    Desired: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
    Reported: 	map[Struct:map[Values:[{Timestamp:1613810501} {Timestamp:1613810501}]] Value:{Timestamp:1613810501}]
    Delta: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
  Version: 	192
  Timestamp: 	1613810501
  ClientToken: 	
delta:
  Value: 	1

> get document
delta:
  Value: 	1
document:
  State: 
    Desired: 	map[Struct:map[Values:[1 2]] Value:1]
    Reported: 	map[Struct:map[Values:[1 2]] Value:2]
    Delta: 	map[Value:1]
  Metadata: 
    Desired: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
    Reported: 	map[Struct:map[Values:[{Timestamp:1613810501} {Timestamp:1613810501}]] Value:{Timestamp:1613810501}]
    Delta: 	map[]
  Version: 	192
  Timestamp: 	1613810502
  ClientToken: 	

> remove one attribute from report
delta:
  Value: 	1
document:
  State: 
    Desired: 	map[Struct:map[Values:[1 2]] Value:1]
    Reported: 	map[Struct:map[Values:[1 2]]]
    Delta: 	map[Value:1]
  Metadata: 
    Desired: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
    Reported: 	map[Struct:map[Values:[{Timestamp:1613810501} {Timestamp:1613810501}]] Value:{Timestamp:1613810504}]
    Delta: 	map[Value:{Timestamp:1613810500}]
  Version: 	193
  Timestamp: 	1613810504
  ClientToken: 	

> add an another attribute to report
document:
  State: 
    Desired: 	map[Struct:map[Values:[1 2]] Value:1]
    Reported: 	map[Struct:map[Values:[1 2]] Value2:1]
    Delta: 	map[Value:1]
  Metadata: 
    Desired: 	map[Struct:map[Values:[{Timestamp:1613810500} {Timestamp:1613810500}]] Value:{Timestamp:1613810500}]
    Reported: 	map[Struct:map[Values:[{Timestamp:1613810501} {Timestamp:1613810501}]] Value:{Timestamp:1613810504} Value2:{Timestamp:1613810505}]
    Delta: 	map[Value:{Timestamp:1613810500}]
  Version: 	194
  Timestamp: 	1613810505
  ClientToken: 	
delta:
  Value: 	1

> delete
document:
```

</details>

**Comment by pablocastilla** [comment]
Sure! I'll do it this Monday

**Comment by pablocastilla** [comment]
it works, now I can do a s.Get(ctx) and I have everything there. Just one question, does this call makes a request to AWS? Couldn't it pass the metadata info to the OnDelta event directly?

**Comment by at-wat** [comment]
`s.Get(ctx)` makes a request.
Latest thing document (including metadata) can be accessed by `s.Document()` without making a new request to AWS.

For example:
```go
s.OnDelta(func(delta map[string]interface{}) {
  val := delta["attribute"]
  meta := s.Document().Metadata.Delta["attribute"].(shadow.Metadata)
  ...
})
```
As I wrote [here](https://github.com/seqsense/aws-iot-device-sdk-go/issues/211#issuecomment-781922139), Thing Document returned by `Shadow.Document()` in `OnDelta` callback is guaranteed to be the version given by the delta event.

**Comment by at-wat** [comment]
BTW, I noticed that AWS IoT ommits Metadata.Delta from Get response, and `s.Document().Metadata.Delta` was cleared when calling `s.Get(ctx)`.

**Comment by pablocastilla** [comment]
perfect! Looking forward for the release, thanks so much for your work :)

**Comment by at-wat** [comment]
@seqsense/cloud_backend please take a look

File: shadow/shadow.go [hunk 2]
```diff
@@ -105,6 +105,8 @@ func (s *shadow) getAccepted(msg *mqtt.Message) {
 		return
 	}
 	s.mu.Lock()
+	// For some reason, AWS IoT ommits Metadata.Delta from Get response. Keep previous Metadata.Delta.
```
Comment (kamatama41) [inline shadow/shadow.go:108]: omits

**Review by kamatama41 — APPROVED** [review]
LGTM 😺

File: shadow/shadow.go (hunk 2 above)
Comment (at-wat) [reply shadow/shadow.go:108]:
```suggestion
	// For some reason, AWS IoT omits Metadata.Delta from Get response. Keep previous Metadata.Delta.
```

**Comment by at-wat** [comment]
Released as v5.0.0-rc1 (waiting there are no more breaking changes)
https://github.com/seqsense/aws-iot-device-sdk-go/releases/tag/v5.0.0-rc1

**Pull request merged**